### PR TITLE
init: cross platform image builder without docker

### PR DESCRIPTION
**Pull request by maisem**
We recently added support for building cross platform docker containers using docker buildx. However, docker does cross platform builds by emulating the target arch using qemu. That is obviously very slow to build and consumes 100% of the cpu, and when run on arm64 machines has a tendency to fail completely.

This in contrast, is a small go binary which uses `GOOS= GOARCH= go build` directly to compile the binaries and then uses `google/go-containerregistry` to create and publish the new containers based on the desired platform.

This is inspired by `google/ko` which is awesome but doesn't support multiple binaries in a single container.

Signed-off-by: Maisem Ali <maisem@tailscale.com>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# `mkctr`: cross platform container builder for go
+
+`mkctr` is a small go binary which uses `GOOS= GOARCH= go build` directly to compile go binaries and then uses [go-containerregistry](https://github.com/google/go-containerregistry) to create and publish the new containers based on the desired platforms.
+
+This is inspired by [ko](https://github.com/google/ko) which is awesome but doesn't support multiple binaries in a single container.
+
+## Usage
+
+```bash
+mkctr \
```
Comment (bradfitz) [inline README.md:10]: where do you specify the GOOS/GOARCH values? show that?

File: main.go [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/empty"
+	"github.com/google/go-containerregistry/pkg/v1/mutate"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/google/go-containerregistry/pkg/v1/tarball"
+	"github.com/google/go-containerregistry/pkg/v1/types"
+	"tailscale.com/types/logger"
```
Comment (bradfitz) [inline main.go:30]: I'm not sure this dep is worth it. I'd just use the standard library log package and logger.New in this codebase.

File: README.md (hunk 1 above)
Comment (maisem) [reply README.md:10]: they are auto discovered based on the provided image, added a note.

File: main.go (hunk 2 above)
Comment (maisem) [reply main.go]: done

File: main.go [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+	desc, err := remote.Get(baseRef, opts...)
+	if err != nil {
+		return nil, err
+	}
+	return desc, nil
+}
+
+func verifyPlatform(p v1.Platform) error {
+	if p.OS != "linux" {
+		return fmt.Errorf("unsupported os: %v", p.OS)
```
Comment (bradfitz) [inline main.go:131]: OS

File: main.go [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+	}
+	defer os.RemoveAll(tmpDir)
+
+	env := append(os.Environ(),
+		"CGO_ENABLED=0",
+		"GOOS="+cf.OS,
+		"GOARCH="+cf.Architecture,
+	)
+	if cf.Architecture == "arm" {
+		env = append(env, "GOARM=5")
```
Comment (bradfitz) [inline main.go:271]: for all ARM or just the old ARM?

I don't know what values cf.Architecture can have.

But we don't want GOARM=5 for e.g. new Raspberry Pis.

File: main.go [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+	}
+	out := f.Name()
+	if err := f.Close(); err != nil {
+		return "", err
+	}
+	cmd := exec.Command(
+		"go",
+		"build",
+		"-v",
+		"-tags=xversion",
```
Comment (bradfitz) [inline main.go:309]: what's xversion for?

File: main.go [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+	return refs, nil
+}
+
+var (
+	baseImage = flag.String("base", "", "base image for container")
+	gopaths   = flag.String("gopaths", "", "comma-separated list of go paths in src:dst form")
+	files     = flag.String("files", "", "comma-separated list of static files in src:dst form")
+	repos     = flag.String("repos", "", "comma-seperated list of image registries")
+	tagFlag   = flag.String("tags", "", "comma-seperated tags")
+	ldFlags   = flag.String("ldflags", "", "ldflags for gopaths")
```
Comment (bradfitz) [inline main.go:81]: what's gopaths?

**Review by bradfitz — APPROVED** [review]

File: main.go (hunk 4 above)
Comment (maisem) [reply main.go]: parsed the arch varient

File: main.go (hunk 5 above)
Comment (maisem) [reply main.go]: i copied it over from the dockerfile, not required. removed

**Pull request merged**